proof/refine/ARM/Tcb_R.thy

```diff
@@ -2791,29 +2791,205 @@ lemma decodeSetSchedParams_wf[wp]:
+  "\<lbrakk>sc_active sc; valid_sched_context' sc' s'; sc_relation sc n sc'\<rbrakk>
+   \<Longrightarrow> rAmount (refillHd sc') = r_amount (refill_hd sc) \<and> rTime (refillHd sc') = r_time (refill_hd sc)"
+  sorry
+
+(* FIXME RT: move to...? *)
+lemma get_sc_released_corres:
+  "corres (=) (active_sc_valid_refills and sc_at sc_ptr) (valid_objs' and sc_at' sc_ptr)
+          (get_sc_released sc_ptr) (scReleased sc_ptr)"
+  unfolding get_sc_released_def scReleased_def scActive_def refillReady_def sc_released_def
+            getCurTime_def
+  apply (clarsimp simp: bind_assoc)
+  apply (rule corres_split[OF _ get_sc_corres, THEN corres_guard_imp])
+      apply clarsimp
+      apply (rule corres_split[OF _ corres_gets_trivial, simplified])
+         apply (rule corres_symb_exec_r[OF _ get_sc_sp'])
+           apply (rename_tac sc sc' n cur_time cur_time' sc2')
+           apply (rule_tac P="active_sc_valid_refills and sc_at_pred ((=) sc) sc_ptr"
+                           and P'="valid_objs' and ko_at' sc' sc_ptr and ko_at' sc2' sc_ptr"
+                           in corres_return[THEN iffD2, rule_format])
+           apply (clarsimp simp: refill_sufficient_def refill_capacity_def refill_ready_def
+                                 kernelWCETTicks_def)
+           apply normalise_obj_at'
+           apply (frule(1) sc_ko_at_valid_objs_valid_sc')
+           apply clarsimp
+           apply (rule iffI; clarsimp)
+            apply (frule(2) refills_heads_equal_active)
+            apply (clarsimp simp: sc_relation_def active_sc_def)
+           apply (prop_tac "sc_active sc")
+            apply (clarsimp simp: sc_relation_def active_sc_def)
+           apply clarsimp
+           apply (intro conjI)
+            apply (frule(2) refills_heads_equal_active)
+            apply (clarsimp simp: sc_relation_def)
+           apply (frule active_sc_valid_refillsE[OF is_sc_active_kh_simp[THEN iffD1],
+                                                 OF is_sc_active_def2[THEN iffD2],
+                                                 rotated])
+            apply (fastforce simp: sc_at_pred_def obj_at_def)
+           apply (prop_tac "cfg_valid_refills (sc_refill_cfg_of sc)")
+            apply (clarsimp simp: sc_at_pred_def obj_at_def valid_refills_def pred_map_def
+                                  sc_refill_cfgs_of_scs_def map_project_def scs_of_kh_def sc_of_def)
+           apply (clarsimp simp: sc_refill_cfg_of_def cfg_valid_refills_def rr_valid_refills_def
+                                 sp_valid_refills_def split: if_splits)
+          apply wpsimp+
+         apply (clarsimp simp: obj_at'_def)
+        apply (clarsimp simp: state_relation_def)
+       apply wpsimp+
+   apply (clarsimp simp: obj_at_def sc_at_pred_n_def)
+  apply clarsimp
+  done
+
+(* FIXME RT: move to...? *)
+crunches scReleased
+  for inv: P
+
+(* FIXME RT: it's weird and sad that this is necessary. Can we replace this with something more
+             generic?  *)
```

Comment:
Can you say a bit more why we need this to start with? Is it because of the scheduling context size being different from other sizes?

Also, are `sc` and `sc’` below functions? If so, that would be rather unconventional and potentially confusing?